understanding/20/keyboard-no-exception.html

```diff
@@ -29,6 +29,28 @@ <h2>Intent of Keyboard (No Exception)</h2>
          Rather, it means that content that uses path-dependent input cannot conform to this
          Success Criterion and therefore cannot meet Guideline 2.1 at Level AAA. 
       </p>
+
+      <div class="note">
+         <p>Platforms and user agents usually have conventions for how web content or applications
+            are controlled with a keyboard interface. As a best practice, content should follow
+            the platform/user agent conventions. However, deviating from these conventions does
+            not fail the normative requirement of this Success Criterion. For instance, buttons
+            that have focus can generally be activated using both the <kbd>Enter</kbd> key and
+            the <kbd>Space</kbd> bar. If a custom button control in a web application instead only
+            reacts to <kbd>Enter</kbd> (or even a completely custom key or key combination), this does <strong>not</strong> fail the Success Criterion.
```

Comment:
Again, I'm concerned about the unqualified statement.
508 has a requirement that accessibility features are documented, and IBM has used that to direct that if someone introduces unconventional keyboard interaction, it must be clearly documented, and that documentation must be keyboard accessible and should be easily discovered, preferably at the point of interaction.

Finally, if a 'custom' button failed to react to Space, it would be surprising to a keyboard user who ordinarily expects that behaviour -- particularly where the component is given the role of button. I would definitely flag that in a review, and I would be flag that as suspicious of other likely problems with the implementation.
```suggestion
            reacts to <kbd>Enter</kbd> (or even a completely custom key or key combination), this does <strong>not</strong> necessarily fail the Success Criterion.
```

Reply:
@mbgower: I agree with you from an accessibility perspective, but the question is whether the normative text of SC 2.1.1 provides these requirements. I'm afraid not, that's why I created: https://github.com/w3c/wcag/issues/872. And the fact that 508 and IBM go beyond the WCAG requirements is great, but it doesn't change the WCAG.

Reply:
No, just to confirm, you are saying that an item that no one but the author can figure out how to operate by keyboard passes 2.1.1? That is a _very_ tight interpretation of the SC.

Reply:
I get what you're saying. But I do not believe anyone should feel confident using such a tight interpretation of the normative text in a court case or with an indignant client. I do not believe it was the intent of the authors of 2.1.1 that keyboard operation was only a hypothetical achievement where someone may discover how to perform a keyboard operation.
You'll note that I have only added 1 word into my suggested change, to convey that there is a lot of gray area. I'd also very much prefer additional language in the Understanding document on best practices.

Reply:
> That is a very tight interpretation of the SC

and yet, the SC doesn't hint at anything other than that. adding that hint in understanding seems to introduce new normative requirements that simply aren't present in the normative text.

by all means, we can add more "however, as a best practice", but let's not suggest an interpretation that isn't actually in the wording of the SC - and it isn't in the wording of the SC precisely because it's soo wooly/undefined, as there's no authoritative/standardised document anywhere that dictates what the standards for web content are. even 'platform' standards don't necessarily apply, as the platform is not web - and, can't make web content that is platform-agnostic dependent on unknown/unknowable (for authors) platform conventions"

Reply:
The problem is that there are people who test accessibility or are dependent on it and therefore interpret SC 2.1.1 in one way, while the software manufacturers, who would then have to fix the problems, often have no desire to do so and argue with the wording of SC 2.1.1, which is the way it is. Because we unfortunately cannot change the wording, I had suggested creating a new SC. Unfortunately, the opportunity was missed. I also have no objection to explaining in the Understanding that SC 2.1.1 does not say how a website must be operated, but that is what it indirectly means. But will all of the WCAG WG agree with this? I'm afraid not.

Reply:
> I do not believe anyone should feel confident using such a tight interpretation of the normative text in a court case or with an indignant client

that would then turn into a "it follows the letter of WCAG, but isn't actually accessible/usable/logical". it's certainly not the first SC that has this problem. luckily, court cases aren't decided *solely* on the absolute baseline of WCAG. as for discussions with clients...

Reply:
And weak WCAG requirements are bad for accessibility. But even worse are unclear WCAG requirements that are interpreted differently by everyone. In this respect, I am always in favor of clarifications in the Understanding, even if they weaken the degree of accessibility.

Reply:
> adding that hint in understanding seems to introduce new normative requirements that simply aren't present in the normative text.

I don't agree with that. It is possible to add language in on challenges with custom keyboard interaction and how it is recommended to document unusual keyboard interaction (including making it discoverable in the UI). There is no reason that needs to be written in a way that is prescriptive so as to be read as normative.

> introduce new normative requirements that simply aren't present in the normative text.

There are **many** SCs that are based on underlying assumptions that are unfortunately not stated. For example
> [Audio description](https://www.w3.org/TR/WCAG20/#audiodescdef) is provided for all [prerecorded](https://www.w3.org/TR/WCAG20/#prerecordeddef) [video](https://www.w3.org/TR/WCAG20/#videodef) content in [synchronized media](https://www.w3.org/TR/WCAG20/#synchronizedmediadef). (Level AA)

All criteria that make similar uses of variations on the word "describe" have a clear assumption that the description is _accurate_. But a tight interpretation of such SCs is that it passes WCAG when there is _any_ description, regardless of its quality or veractiy. But I would suggest to you that it is in no one's interest to put the following into the SC:

> a description that does not match the visual content is not a failure of this SC

I see what is being proposed with these additions as being along the same lines. Adding the proposed content to the Understanding document without any additional explanation or caution is going to make it easier for someone anyone to point to the WCAG guidance to defend an interaction that is effectively unusable.

Reply:
adding "necessarily" to this PR does not help clarify the interpretation of the SC for auditors and simply codifies the handwavy status quo of the current SC ... which then makes me wonder what the point would be of doing it in the first place. we either want to bring clarity, or we're happy with different auditors coming to different conclusions here

Reply:
How about:
> "Platforms and user agents usually have conventions for how web content or applications
>  are controlled with a keyboard interface. Any method of keyboard will pass the criterion. However, if content does not follow the platform/user agent conventions it
> will be diffiult to use, and testers may assume that it is failing the criterion.
> For instance, buttons that have focus can generally be activated using both the <kbd>Enter</kbd> key and the <kbd>Space</kbd> bar. If a custom button control in a web application instead only reacts to <kbd>Enter</kbd> (or even a completely custom key or key combination), this does <strong>not</strong> fail the Success Criterion."

Reply:
Grafted some of the suggested wording back into my PR here - though I removed the explicit mention of "issues when testing" as they're just as applicable to all regular users as well.